### PR TITLE
Change enumerations to yield

### LINKED ISSUE #552: Proposal: Replace Enumerator objects by yield
In recent analysis of enumerations in this library, I came across [these articles](https://devblogs.microsoft.com/oldnewthing/20080812-00/?p=21273). 

I think that the library would benefit both in performance and code complexity from removing the enumerator objects and replacing them with yield based statements in the corresponding enumeration objects.

This would solve the problem of either having non-readonly members for cursor number and current object or single item arrays (like it is currently) which have allocation costs. Sidenote: The original library (cactoos) opted for non readonly members, breaking the EO rules.

Note: A change should wait for a decision about caching. If caching by default is still preferred, the react 'MemoizedBuffer' object for buffering could be taken as blueprint for an efficient caching mechanism.

Example

```csharp
    public sealed class Filtered<T> : IEnumerable<T>
    {
        /// <summary>
        /// A filtered <see cref="IEnumerable{T}"/> which filters by the given condition <see cref="Func{In, Out}"/>.
        /// </summary>
        /// <param name="src">enumerable to filter</param>
        /// <param name="fnc">filter function</param>
        /// <param name="live">live or sticky</param>
        public Filtered(Func<T, Boolean> fnc, IEnumerable<T> src, bool live)
        {
            this.src = src;
            this.fnc = fnc;
        }

        public IEnumerator<T> GetEnumerator()
        {
            foreach(var item in this.src)
            {
                if (fnc.Invoke(item))
                {
                    yield return item;
                }
            }
            yield break;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}
```

```csharp
    public sealed class Joined<T> : IEnumerable<T>
    {
        /// <summary>
        /// Multiple <see cref="IEnumerable{T}"/> Joined2 together.
        /// </summary>
        /// <param name="items">enumerables to join</param>
        public Joined(IEnumerable<IEnumerable<T>> items)
        {
            this.enumerables = items;
        }

        public IEnumerator<T> GetEnumerator()
        {
            foreach(var enumerable in this.enumerables)
            {
                foreach(var item in enumerable)
                {
                    yield return item;
                }
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}
```

```csharp
    /// <summary>
    /// Enumeration of a single item.
    /// </summary>
    public class Single<T> : IEnumerable<T>
    {
        private readonly T item;

        /// <summary>
        /// Enumeration of a single item.
        /// </summary>
        public Single(T item)
        {
            this.item = item;
        }

        public IEnumerator<T> GetEnumerator()
        {
            yield return this.item;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
```

```csharp
    /// <summary>
    /// Enumeration from array of items.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public sealed class Params<T> : IEnumerable<T>
    {
        private readonly T[] items;

        /// <summary>
        /// Enumeration from array of items.
        /// </summary>
        /// <param name="items"></param>
        public Params(params T[] items)
        {
            this.items = items;
        }

        public IEnumerator<T> GetEnumerator()
        {
            foreach(var item in this.items)
            {
                yield return item;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
```

### LINKED ISSUE #553: Refactor: ManyEnvelope unnecessarily copies all items
### Expected Behavior
ManyEnvelope (live=false mode) remembers only items which have been visited.

### Actual Behavior
The primary ctor copies each item although it might not be needed, consuming unnecessary computation power.

### The log given by the failure.

### LINKED ISSUE #554: Feat: Ternary Enumerable
### Expected Behavior
There is a ternary object which enumerates one of two given enumerables, depending on a given condition.

### Actual Behavior
No such feature.

### The log given by the failure.

### LINKED ISSUE #559: Remove Enumerable's Sticky behaviour where it is unnecessary
### Expected Behavior
Enumerables which do not profit from being sticky in any way do not build an internal cache.

### Actual Behavior
All Enumerables are based on ManyEnumerable, which is stcky by default. It copies all items which it receives while accessing them. If multiple enumerables are stacked together by using their decorator-ctors, it procudes unnecessary caches, copying all items, although only the most shallow cache is finally used. 

### Mention any other details that might be useful
To counter this, the cache should be revised and the Atoms enumerables should offer a live option for all their ctors. Many enumerables do not need any cache at all (think of 'repeated' for an obvious example), and therefore the cache there should be removed.

### The log given by the failure.

### PR DESCRIPTION
**Pull request by MSE1188**
### Please check if the PR fulfills these requirements
- [x] I made sure that my code builds
- [x] I merged the master into this branch before pushing
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?** (check one with "x")
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

### What is the current behavior? (You can also link to an open issue here)
closes #552 
closes #554 
closes #553 
closes #559 

### Does this PR introduce a breaking change? (check one with "x")
- [ ] Yes
- [x] No

###  Other information

### CODE REVIEW

**Comment by Meerownymous** [comment]
Union might be the last object to get working.
Changed to hashtable with comparer, but currently it returns the wrong item (right) where the left is needed.

**Comment by MSE1188** [comment]
thanks a lot, I will take a look into this the next days

**Comment by Meerownymous** [comment]
No problem, was  fun to improve this. I finished the last two enumerables.
I changed the ctors of all objects that I could find to using live objects instead of sticky ones.

I see room for improvement on using the live objects, maybe by introducing shorthand ctors (like the "New" ones) to make them more visible.

File: src/Yaapii.Atoms/Map/Joined.cs [hunk 1]
```diff
@@ -64,24 +64,26 @@ public sealed class Joined : MapEnvelope
...
+                                new Live<IEnumerator<IKvp>>(() =>
                                 {
-                                    list = new Enumerable.Joined<IKvp>(list, new KvpOf(key, () => dict[key]));
-                                }
-                                return list.GetEnumerator();
-                            })
+                                    IEnumerable<IKvp> list = new ManyOf<IKvp>();
+                                    foreach (var key in dict.Keys)
+                                    {
+                                        list = new Enumerable.Joined<IKvp>(list, new KvpOf(key, () => dict[key]));
```
Comment (MSE1188) [inline src/Yaapii.Atoms/Map/Joined.cs:75]: I do not like this as still means, that every dict entry would be mapped into a joined decorator...

Imagine joining 2 dicts with 100.000 entries each (which is not a rare thing). The Joined-decorations would stack up accordingly and would probably result in a stack-overflow as described in the issues.
I already made a PR targeting this class specifically: https://github.com/icarus-consulting/Yaapii.Atoms/pull/542

File: src/Yaapii.Atoms/Enumerable/Single.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Yaapii.Atoms.Enumerable
+{
+    /// <summary>
+    /// Enumeration of a single item.
+    /// </summary>
+    public class Single<T> : IEnumerable<T>
```
Comment (MSE1188) [inline src/Yaapii.Atoms/Enumerable/Single.cs:10]: I like this one :)

File: src/Yaapii.Atoms/Map/MapOf.cs [hunk 3]
```diff
@@ -78,7 +81,8 @@ public sealed class MapOf : MapEnvelope
         public MapOf(IKvp entry, params IKvp[] more) : base(() =>
             new LazyDict(
                 new Enumerable.Joined<IKvp>(
-                    new ManyOf<IKvp>(entry),
+                    live: true,
+                    new LiveMany<IKvp>(entry),
```
Comment (MSE1188) [inline src/Yaapii.Atoms/Map/MapOf.cs:85]: You could use it (```new Single<IKvp>```) here I guess

**Review by MSE1188 — COMMENTED** [review]
Sorry, that it took so long. There are a lot of other topics going on currently.. As I requested this PR I can neither approve nor request changes on this PR so this is only a comment

File: src/Yaapii.Atoms/Enumerable/EnumeratorAsEnumerable.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            this.enumerator.Reset();
+            while (this.enumerator.MoveNext())
+            {
+                yield return this.enumerator.Current;
+            }
+            yield break;
```
[the remark this answers is not in the document]
Comment (Meerownymous) [reply src/Yaapii.Atoms/Enumerable/EnumeratorAsEnumerable.cs:36]: Thats right. Seems I wasn't familiar then either :)

File: src/Yaapii.Atoms/Map/MapOf.cs (hunk 3 above)
Comment (Meerownymous) [reply src/Yaapii.Atoms/Map/MapOf.cs]: Aye

File: src/Yaapii.Atoms/Enumerable/Single.cs (hunk 2 above)
Comment (Meerownymous) [reply src/Yaapii.Atoms/Enumerable/Single.cs]: It says it likes you too

**Comment by Meerownymous** [comment]
@MSE1188 I did not get why you could not approve - but I cannot merge without approval :)

File: src/Yaapii.Atoms/Map/Joined.cs (hunk 1 above)
Comment (Meerownymous) [reply src/Yaapii.Atoms/Map/Joined.cs]: You might want to have a look at [this suggestion](https://github.com/icarus-consulting/Yaapii.Atoms/issues/562)

**Review by MSE1188 — COMMENTED** [review]
seems good. I can merge this on my own though

**Pull request merged**